File: leds_unity/Assets/Trail.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class Trail
{
    Color color;
    public List<Particle> all;

    public void Init()
    {
        all = new List<Particle>();
    }
    int lastLedID;
    public void OnUpdate(float deltaTime, int ledID, float _fade_desaceleration)
    {
        foreach (Particle p in all)
            p.OnUpdate(deltaTime, _fade_desaceleration);

        if (lastLedID == ledID) return;
        Particle trail = GetParticleOrCreate();
        trail.Init(ledID, 180);
    }
    Particle GetParticleOrCreate()
    {
        foreach (Particle t in all)
            if (t.IsAvailable())
                return t;
        Particle trail = new Particle();
        all.Add(trail);
        return trail;
    }
}

File: micropython/agosto_2023/trail.py
import TrailParticle, utils


class Trail:
    color = 2
    all = []
    lastLedID = 0
    trailSpeed = 30
    global leds
    def Init(self, _color : int,  _totalParticles : int, leds) :
        self.leds = leds
        self.color = _color
        self.all = []
        
        for a in range(_totalParticles):   
            particle = TrailParticle.TrailParticle()
            self.all.append(particle)
                
    def Move(self, ledID, _speed, deltaTime):        
        lN = len(self.all)      
        for a in range(lN):        
            p = self.all[a]
            if p.IsAvailable() == False:
                p.OnUpdate(deltaTime*self.trailSpeed)
                
                self.leds.AddAlpha(self.color, p.ledID, p.value)
        if _speed<0:
            ledID = ledID+1
        else:
            ledID = ledID-1
        
        if ledID != self.lastLedID:
            self.lastLedID = ledID 
            for a in range(lN):        
                p = self.all[a]
                if p.IsAvailable():
                    p.Init(ledID)
                    return


